[WPE] Generate damage region information

aperezdc opened the pull request:
#### a6a10cdfb8baf3512b99b36741595efa8c6feb06
<pre>
[WPE] Generate damage region information
<a href="https://bugs.webkit.org/show_bug.cgi?id=265777">https://bugs.webkit.org/show_bug.cgi?id=265777</a>

Reviewed by Carlos Garcia Campos.

Includes code and ideas from Lauro Moura, Pawel Lampe, Carlos Garcia
Campos, and Miguel Gomez.

Implement tracking which part of a frame change since the previous one.
This may be used to let display systems optimize presentation of the
rendered frames by updating only the parts that have been &quot;damaged&quot;.
While usually negligible on desktops, this can result in significant
savings on embedded devices, particularly in memory bandwith and GPU
usage.

Information about damaged frames is collected mainly from repainted
rectangles (from CoordinatedGraphicsLayer) and animated layers (from
TextureMapperLayer), and recorded in a new helper Damage class.

This is a first pass that only propagates the recorded Damage along with
DMA-BUF buffers, and forwards the information to the Wayland compositor
support in WPEPlatform. It should be possible to propagate the
information through EGL where the EGL_{KHR,EXT}_swap_buffers_with_damage
extension is present, which would work when the WPEPlatform API is not
being used; gdk_dmabuf_texture_builder_set_update_region() could be
used for the GTK port; and the optional FB_DAMAGE_CLIPS property for
the DRM platform. These are left to be implemented later on.

The intention for now is to lay out the foundations, and hence the
feature is gated behind a runtime flag. Being an initial implementation,
there are number of shortcomings that may be addressed as follow-ups.

There is an attempt at detecting whether 3D transforms are used, with
the intention of invalidating the whole frame; this mostly works but it
would be better to properly handle these instead of causing full-frame
damage.

There is a pre-existing issue which manifests itself more obviously with
damage tracking enabled: while animating, the renderer content may not
be as crisp as static; typically WebKit would repaint the crisper
version of the content after animations have completed, but if some
parts have not changed since the last frame, they will show up fuzzy
because they are not considered as damaged. This is independent from
damage tracking.

Enabling damage tracking seems add negligible CPU usage: enabling it is
neutral on Speedometer and MotionMark, which for the WPE port are CPU
bound due to rasterization through Cairo being CPU bound.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml: Add feature
  flags to enable damage tracking, and to produce a single rectangle
  with the bounds of the damaged region instead of multiple rectangles.
* Source/WebCore/Headers.cmake:
* Source/WebCore/platform/graphics/Damage.h: Added.
(WebCore::Damage::invalid):
(WebCore::Damage::region const):
(WebCore::Damage::bounds const):
(WebCore::Damage::rects const):
(WebCore::Damage::isEmpty const):
(WebCore::Damage::isInvalid const):
(WebCore::Damage::invalidate):
(WebCore::Damage::add):
(WebCore::Damage::mergeIfNeeded):
(WebCore::Damage::Damage):
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::markDamageRectsUnreliable):
* Source/WebCore/platform/graphics/nicosia/NicosiaCompositionLayer.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::TextureMapperLayer):
(WebCore::TextureMapperLayer::paintSelf):
(WebCore::TextureMapperLayer::addChild):
(WebCore::TextureMapperLayer::removeFromParent):
(WebCore::TextureMapperLayer::removeAllChildren):
(WebCore::TextureMapperLayer::applyAnimationsRecursively):
(WebCore::TextureMapperLayer::acceptDamageVisitor):
(WebCore::TextureMapperLayer::dismissDamageVisitor):
(WebCore::TextureMapperLayer::recordDamage):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
(WebCore::TextureMapperLayer::clearDamage):
(WebCore::TextureMapperLayer::invalidateDamage):
(WebCore::TextureMapperLayer::addDamage):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.cpp:
(WebCore::CoordinatedGraphicsLayer::setContentsOpaque):
(WebCore::CoordinatedGraphicsLayer::markDamageRectsUnreliable):
(WebCore::CoordinatedGraphicsLayer::setNeedsDisplay):
(WebCore::CoordinatedGraphicsLayer::setNeedsDisplayInRect):
(WebCore::CoordinatedGraphicsLayer::flushCompositingStateForThisLayerOnly):
(WebCore::CoordinatedGraphicsLayer::computeTransformedVisibleRect):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::layerRendererStyleHas3DTransformOperation):
(WebCore::RenderLayerBacking::updateAfterDescendants):
* Source/WebKit/Shared/CoordinatedGraphics/CoordinatedGraphicsScene.cpp:
(WebKit::CoordinatedGraphicsScene::CoordinatedGraphicsScene):
(WebKit::CoordinatedGraphicsScene::paintToCurrentGLContext):
(WebKit::texmapLayer):
(WebKit::CoordinatedGraphicsScene::updateSceneState):
(WebKit::CoordinatedGraphicsScene::ensureRootLayer):
(WebKit::CoordinatedGraphicsScene::purgeGLResources):
(WebKit::CoordinatedGraphicsScene::recordDamage):
* Source/WebKit/Shared/CoordinatedGraphics/CoordinatedGraphicsScene.h:
(WebKit::CoordinatedGraphicsScene::lastDamage const):
* Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.cpp:
(WebKit::ThreadedCompositor::create):
(WebKit::ThreadedCompositor::ThreadedCompositor):
(WebKit::ThreadedCompositor::renderLayerTree):
* Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.h:
* Source/WebKit/UIProcess/dmabuf/AcceleratedBackingStoreDMABuf.messages.in:
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp:
(WebKit::AcceleratedBackingStoreDMABuf::frame):
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.h:
* Source/WebKit/UIProcess/wpe/AcceleratedBackingStoreDMABuf.cpp:
(WebKit::AcceleratedBackingStoreDMABuf::frame):
* Source/WebKit/UIProcess/wpe/AcceleratedBackingStoreDMABuf.h:
* Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
(wpe_view_render_buffer):
* Source/WebKit/WPEPlatform/wpe/WPEView.h:
* Source/WebKit/WPEPlatform/wpe/drm/WPEViewDRM.cpp:
(wpeViewDRMRenderBuffer):
* Source/WebKit/WPEPlatform/wpe/headless/WPEViewHeadless.cpp:
(wpeViewHeadlessRenderBuffer):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
* Source/WebKit/WebProcess/WebPage/AcceleratedSurface.h:
(WebKit::AcceleratedSurface::didRenderFrame):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.cpp:
(WebKit::LayerTreeHost::LayerTreeHost):
(WebKit::LayerTreeHost::scrollNonCompositedContents):
(WebKit::LayerTreeHost::didChangeViewport):
(WebKit::LayerTreeHost::didRenderFrame):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.h:
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
(WebKit::AcceleratedSurfaceDMABuf::didRenderFrame):
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.h:
* Source/WebKit/WebProcess/WebPage/libwpe/AcceleratedSurfaceLibWPE.cpp:
(WebKit::AcceleratedSurfaceLibWPE::didRenderFrame):
* Source/WebKit/WebProcess/WebPage/libwpe/AcceleratedSurfaceLibWPE.h:

Canonical link: <a href="https://commits.webkit.org/280193@main">https://commits.webkit.org/280193@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b9469c9579ca5dfc1472b3293248ac1707e6e73e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55966 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35290 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8434 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58965 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6398 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58092 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42911 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6594 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/45071 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4423 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57995 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33184 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48258 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26209 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29967 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5586 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4541 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/49042 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/51943 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5857 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60552 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/55202 "Built successfully and passed tests") | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-1-Build-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/5977 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/52499 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48327 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/52026 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/76963 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8275 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31119 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/76963 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32203 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33285 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31951 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -59,7 +59,10 @@ struct TextureMapperLayer::ComputeTransformData {
+TextureMapperLayer::TextureMapperLayer(const Settings& settings)
+    : m_settings(settings)
+{
+}
```
This is a layering violation, platform should not have access to Settings from WebCore, we need a way to pass the two settings needed as bool or flags or whatever we want.

aperezdc replied:
Fixed, the `WebCore::Settings` reference is no longer passed down to `TextureMapperLayer`.

carlosgcampos commented on Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.h:
```diff
@@ -49,6 +53,8 @@ class ThreadedCompositor : public CoordinatedGraphicsSceneClient, public ThreadS
     class Client {
     public:
+        virtual const WebCore::Settings& settings() = 0;
+
```
I'm not sure about the settings to be honest, do we really need them, how are we going to expose them? But if we really need settings, I don't think we should be passing the Settings object this way. We can add an enum with flags with the settings we need and pass that instead.

aperezdc replied:
Done, I went with an enum that has three posible values.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -631,7 +630,13 @@ static gboolean wpeViewWaylandRenderBuffer(WPEView* view, WPEBuffer* buffer, GEr
+        for (unsigned i = 0; i < nDamageRects; ++i)
+            wl_surface_damage_buffer(wlSurface, damageRects[i].x, damageRects[i].y, damageRects[i].width, damageRects[i].height);
+    } else
+        wl_surface_damage(wlSurface, 0, 0, INT32_MAX, INT32_MAX);
```
why INT32_MAX?

aperezdc replied:
It is a common idiom to mean “damage the whole surface”, I've seen it used plenty in the wild and in documentation, e.g. [in the Wayland Book](https://wayland-book.com/surfaces-in-depth/damaging-surfaces.html).

psaavedra commented on Source/WebCore/Headers.cmake:
```diff
@@ -1886,6 +1886,7 @@ set(WebCore_PRIVATE_FRAMEWORK_HEADERS
     platform/graphics/ColorUtilities.h
     platform/graphics/ComplexTextController.h
     platform/graphics/ContentTypeUtilities.h
+    platform/graphics/Damage.h
```
I dont see this file in the patch. Where does this come from?

psaavedra replied:
fixed by @aperezdc 
 forcing a new push:
 https://github.com/WebKit/WebKit/pull/27313/commits/bdc57d9e3e7c37750afe8690a5344f673b9ba41f

aperezdc replied:
Removed.

carlosgcampos commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -0,0 +1,117 @@
+public:
+    using Rects = Vector<IntRect, 1>;
+
+    explicit Damage() = default;
```
explicit is not needed here, no?

carlosgcampos commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    ALWAYS_INLINE const Region& region() const { return m_region; }
+    ALWAYS_INLINE IntRect bounds() const { return m_region.bounds(); }
+    ALWAYS_INLINE Rects rects() const { return m_region.rects(); }
```
Since Region is exposed, I guess caller can do damage.region().foo() no?

aperezdc replied:
Not really needed, mostly a matter of taste. I'll remove it.

aperezdc replied:
Done.

carlosgcampos commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -0,0 +1,117 @@
+    ALWAYS_INLINE const Region& region() const { return m_region; }
+    ALWAYS_INLINE IntRect bounds() const { return m_region.bounds(); }
+    ALWAYS_INLINE Rects rects() const { return m_region.rects(); }
+    ALWAYS_INLINE bool isEmpty() const  { return !m_invalid && m_region.isEmpty(); }
```
Should we check it's valid in the other getters? what's exactly and invalid damage?

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -218,8 +221,31 @@ void TextureMapperLayer::paintSelf(TextureMapperPaintOptions& options)
+        // been covered by the damage tracking in setNeedsDisplay/setNeedsDisplayInRect
+        // calls from CoordinatedGraphicsLayer.
+        if (m_recordDamage) {
+            if (m_damage.isInvalid())
+                recordDamage(layerRect(), transform, options);
```
So, invalid damage is used to damage the whole layer right? Would it be possible to create a Damage with a Region that has a single rect with max int values as size? That way we don't need to handle that as a special case.

aperezdc replied:
Whenenver we find a situation in which we need to damage the *whole* frame (and not only the current layer!) then we mark a `Damage` object as invalid. Then when damage is propagated to the scene `Damage` any invalid layer in the tree will cause the scene one to also be marked invalid. At the point where we mark layer damage as invalid, we may not know the exact extents because a layer that is 3D-transformed may not cover the area corresponding to its size!

The way I like to think about it is: if any `Damage` is marked invalid, it will “taint” and damage any other `Damage` where it gets combined, resulting in the full frame considered damaged.

Hopefully we may be able to remove the concept of invalidation and do as you suggest in all cases in the future. For example this patch already does what you suggest for non-transformed content layers: the whole layer rect is damaged instead.

Should I add a comment about how this works to the `Damage` class source?

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
```diff
@@ -40,16 +40,20 @@ template<> struct IsDeprecatedWeakRefSmartPointerException<WebCore::TextureMappe
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    TextureMapperLayer();
+    TextureMapperLayer(bool recordDamage);
```
explicit. I would also add an enum class instead of passing a bool

aperezdc replied:
Done!

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
```diff
@@ -40,16 +40,20 @@ template<> struct IsDeprecatedWeakRefSmartPointerException<WebCore::TextureMappe
 
+class TextureMapperLayerDamageVisitor {
+public:
+    virtual void recordDamage(FloatRect) = 0;
```
```suggestion
    virtual void recordDamage(const FloatRect&) = 0;
```

aperezdc replied:
Done.

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -218,8 +221,31 @@ void TextureMapperLayer::paintSelf(TextureMapperPaintOptions& options)
+            if (m_damage.isInvalid())
+                recordDamage(layerRect(), transform, options);
+            else {
+                for (auto& rect : m_damage.rects()) {
```
```suggestion
                for (const auto& rect : m_damage.rects()) {
```

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
```diff
@@ -118,6 +122,14 @@ class WEBCORE_EXPORT TextureMapperLayer : public CanMakeWeakPtr<TextureMapperLay
+    void acceptDamageVisitor(TextureMapperLayerDamageVisitor&);
+    void dismissDamageVisitor();
+
+    ALWAYS_INLINE void clearDamaged();
```
```suggestion
    ALWAYS_INLINE void clearDamage();
```

aperezdc replied:
Done.

aperezdc replied:
Thanks for the nudge, this suggestion actually simplified code a bit.

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
```diff
@@ -242,6 +255,10 @@ class WEBCORE_EXPORT TextureMapperLayer : public CanMakeWeakPtr<TextureMapperLay
 #endif
     bool m_isBackdrop { false };
     bool m_isReplica { false };
+    bool m_recordDamage { false };
```
we call `recordDamage` to the method to record a damage, I would call this `m_shouldRecordDamage`

aperezdc replied:
Better, I'll use the same `enum class` instead you suggested above.

aperezdc replied:
Done, using the newly introduced `enum class DamagePropagation` type.

aperezdc replied:
This ended up like this with the suggested change:

```cpp
    Vector<WebCore::IntRect, 1> damageRects;
    if (damage) {
        if (damage->isEmpty())
            damageRects.append({ 0, 0, 0, 0 });
        else
            damageRects = damage->rects();
    }
```

...which is, arguably, easier to understand. Thanks for this suggestion.

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.cpp:
```diff
@@ -1051,6 +1063,17 @@ void CoordinatedGraphicsLayer::flushCompositingStateForThisLayerOnly()
                     state.eventRegion = eventRegion();
+                if (localDelta.damageChanged) {
+                    state.damage = Damage();
+                    if (m_needsDisplay.completeLayer)
+                        state.damage.add(FloatRect({ }, m_size));
+                    else {
+                        for (const auto& rect : m_needsDisplay.rects)
+                            state.damage.add(rect);
+                    }
+                    if (m_damagedRectsAreUnreliable)
+                        state.damage.invalidate();
```
if `m_damagedRectsAreUnreliable` is true we set the damage to invalidate it later. If we handle invalidation as damage full layer, maybe we can do:

```
if (m_needsDisplay.completeLayer || m_damagedRectsAreUnreliable)
```

aperezdc replied:
Read my comment above about this: if the layer has 3D transformations it may cover a different part of the frame than the 2D rectangle that matches its size. OTOH, your suggestion made me realize that we can shortcut some work this way:

```suggestion
                    if (m_damagedRectsAreUnreliable)
                        state.damage.invalidate();
                    else if (m_needsDisplay.completeLayer)
                        state.damage.add(FloatRect({ }, m_size));
                    else {
                        for (const auto& rect : m_needsDisplay.rects)
                            state.damage.add(rect);
                    }
```

That is: if we need to consider the whole frame damaged, just invalidate the `Damage` instance instead of first adding rects to it and then invalidate it as well. WDYT?

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.cpp:
```diff
@@ -1359,7 +1382,8 @@ void CoordinatedGraphicsLayer::computeTransformedVisibleRect()
     m_layerTransform.setChildrenTransform(childrenTransform());
     m_layerTransform.combineTransforms(parent() ? downcast<CoordinatedGraphicsLayer>(*parent()).m_layerTransform.combinedForChildren() : TransformationMatrix());
 
-    m_cachedInverseTransform = m_layerTransform.combined().inverse().value_or(TransformationMatrix());
+    m_cachedCombinedTransform = m_layerTransform.combined();
+    m_cachedInverseTransform = m_cachedCombinedTransform.inverse().value_or(TransformationMatrix());
```
Why do we need to split this?

aperezdc replied:
I think this was done at some point while trying to figure out inverse 3D transforms, but I am not sure at this point anymore. Shouldn't be needed with the patch in its current form, so I can revert this bit before landing.

carlosgcampos commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -1705,6 +1717,10 @@ void RenderLayerBacking::updateAfterDescendants()
 
+    if (layerRendererStyleHas3DTransformOperation(m_owningLayer)
+        || m_owningLayer.hasCompositedScrollableOverflow())
+        m_graphicsLayer->markDamageRectsUnreliable();
```
So, what we want is to make sure that the whole layer is damaged, no? What about using setShouldDamageFullLayer or something similar instead of marking damage rects as unreliable?

aperezdc replied:
The whole *frame* needs to be invalidated, not the layer being checked.

carlosgcampos commented on Source/WebKit/Shared/CoordinatedGraphics/CoordinatedGraphicsScene.cpp:
```diff
@@ -37,8 +37,9 @@
 using namespace WebCore;
 
-CoordinatedGraphicsScene::CoordinatedGraphicsScene(CoordinatedGraphicsSceneClient* client)
+CoordinatedGraphicsScene::CoordinatedGraphicsScene(CoordinatedGraphicsSceneClient* client, bool recordDamage)
```
Same here, better to use an enum and rename it as shouldRecordDamage

aperezdc replied:
Done.

carlosgcampos commented on Source/WebKit/Shared/CoordinatedGraphics/CoordinatedGraphicsScene.cpp:
```diff
@@ -466,6 +480,11 @@ void CoordinatedGraphicsScene::detach()
     m_client = nullptr;
 }
 
+void CoordinatedGraphicsScene::recordDamage(FloatRect damagedRect)
```
```suggestion
void CoordinatedGraphicsScene::recordDamage(const FloatRect& damagedRect)
```

aperezdc replied:
Done.

carlosgcampos commented on Source/WebKit/Shared/CoordinatedGraphics/CoordinatedGraphicsScene.h:
```diff
@@ -70,6 +71,9 @@ class CoordinatedGraphicsScene : public ThreadSafeRefCounted<CoordinatedGraphics
     void setActive(bool active) { m_isActive = active; }
 
+    const WebCore::Damage& lastDamage() const { return m_damage; }
+    void recordDamage(WebCore::FloatRect) override;
```
```suggestion
    void recordDamage(const WebCore::FloatRect&) override;
```

aperezdc replied:
Done.

carlosgcampos commented on Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.cpp:
```diff
@@ -103,7 +103,7 @@ ThreadedCompositor::ThreadedCompositor(Client& client, ThreadedDisplayRefreshMon
 #endif
 
-        m_scene = adoptRef(new CoordinatedGraphicsScene(this));
+        m_scene = adoptRef(new CoordinatedGraphicsScene(this, m_client.settings().propagateDamagingInformation()));
```
I think we should pass the setting directly as construct parameter instead of exposing all the settings in the ThreadedCompositor client

carlosgcampos commented on Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.cpp:
```diff
@@ -277,10 +277,22 @@ void ThreadedCompositor::renderLayerTree()
+    const auto& frameDamage = ([this, &boundsDamage]() -> const WebCore::Damage& {
+        const auto& damage = m_scene->lastDamage();
+        if (m_client.settings().propagateDamagingInformation() && !damage.isInvalid()) {
+            if (m_client.settings().unifyDamagedRegions()) {
```
Same for this setting, maybe we can unify both in a DamageFlags OptionSet that the threaded compositor receives at construction. I don't think those settings are expected to change at runtime, right?

aperezdc replied:
They may change at runtime. For example in the MiniBrowser from the settings page (or in Ephy when I manage to update and land the patch to add the feature flags to the settings dialog). Being able to toggle them has been useful at times during development.

carlosgcampos commented on Source/WebKit/UIProcess/dmabuf/AcceleratedBackingStoreDMABuf.messages.in:
```diff
@@ -24,5 +24,5 @@ messages -> AcceleratedBackingStoreDMABuf NotRefCounted {
     DidCreateBufferSHM(uint64_t id, WebCore::ShareableBitmapHandle handle)
     DidDestroyBuffer(uint64_t id)
-    Frame(uint64_t id)
+    Frame(uint64_t id, WebCore::Damage damage)
```
I think we could just pass a `std::optional<WebCore::Region>` and we don't need a serializer for the damage.

aperezdc replied:
Will do, and remove the `Damage.serialization.in` file along with the change.

carlosgcampos commented on Source/WebKit/UIProcess/wpe/AcceleratedBackingStoreDMABuf.cpp:
```diff
@@ -133,9 +133,28 @@ void AcceleratedBackingStoreDMABuf::frame(uint64_t bufferID)
+        ASSERT(damageRects.size() <= std::numeric_limits<guint>::max());
+        rects = reinterpret_cast<const WPERectangle*>(damageRects.data());
+        nRects = damageRects.size();
+    }
```
I think we could use Vector<IntRect, 1> for all the cases. When invalid we don't need to do anything, when empty just append({ }).

carlosgcampos commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -0,0 +1,122 @@
+enum class DamagePropagation : bool {
+    No,
+    Yes,
+};
```
Any reason not to make this an enum of the Damage class instead? It would be `Damage::Propagation` or maybe `Damage::ShouldPropagate` is clearer.

aperezdc replied:
Then we would need to `#include <WebCore/Damage.h>` whenever we want to use it because as a nested type it cannot be forward-declared. My reason to do this was to avoid needing to always have the header included, but double-checking things it seems that places that use this `enum` are already including it... so no harm in nesting it. I'll change it.

carlosgcampos commented on Source/WebKit/Platform/Logging.h:
```diff
@@ -51,6 +51,7 @@ extern "C" {
     M(CacheStorage) \
     M(ContentObservation) \
     M(ContextMenu) \
+    M(CoordinatedGraphics) \
```
Are we really using this new logging channel in the end?

aperezdc replied:
No, it's a leftover from an older version of the patch. I'll remove that.

carlosgcampos commented on Source/WebKit/Shared/CoordinatedGraphics/CoordinatedGraphicsScene.h:
```diff
@@ -53,9 +53,10 @@ class CoordinatedGraphicsSceneClient {
+    , public WebCore::TextureMapperLayerDamageVisitor {
 public:
-    explicit CoordinatedGraphicsScene(CoordinatedGraphicsSceneClient*);
+    explicit CoordinatedGraphicsScene(CoordinatedGraphicsSceneClient*, WebCore::DamagePropagation);
```
explicit is no longer needed.

aperezdc replied:
Right, I will remove it.

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.cpp:
```diff
@@ -369,6 +373,11 @@ void LayerTreeHost::frameComplete()
+const WebCore::Settings& LayerTreeHost::settings()
+{
+    return m_webPage.corePage()->settings();
+}
```
Are Settings thread safe?

aperezdc replied:
Apparently not. I guess passing the two flags so they are copied, but then we lose the ability to flip the settings at runtime. Any suggestion on how to go about this?

carlosgcampos replied:
I would use an enum for the two flags, that is passed to the threaded compositor constructor. Then you can define a function ot be called when the settings change with `webcoreOnChange: functionName` in UnifiedWebPreferences.yaml, it could be the same function for both settings if we use the same flags enum. So, when it changes the threaded compositor is notified about the new values.

aperezdc replied:
I didn't bother with this in the end. I think it's okay if only the new web views that are created after changing the settings get the new configuration.

carlosgcampos commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -5231,6 +5231,20 @@ ProcessSwapOnCrossSiteNavigationEnabled:
       default: true
 
+PropagateDamagingInformation:
+   type: bool
+   status: testable
+   category: dom
+   humanReadableName: "Propagate Damaging Information"
+   humanReadableDescription: "Propagate Damaging Information"
```
We should add `condition: USE(COORDINATED_GRAPHICS)` or platform GTK/WPE, but this setting is not used by other ports, no?

aperezdc replied:
I think gating it on `COORDINATED_GRAPHICS` makes sense, I'll add that.

aperezdc replied:
Done.

carlosgcampos commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -6796,6 +6810,20 @@ UnifiedPDFEnabled:
       default: false
 
+UnifyDamagedRegions:
+   type: bool
+   status: testable
+   category: dom
+   humanReadableName: "Unify Damaged Regions"
+   humanReadableDescription: "Unify Damaged Regions"
```
Same for this setting, it should be conditional.

carlosgcampos commented on Source/WebCore/platform/graphics/Damage.h:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    explicit Damage(bool invalid, Region&& region = { })
+        : m_invalid(invalid), m_region(WTFMove(region)) { }
```
```suggestion
        : m_invalid(invalid)
        , m_region(WTFMove(region)) 
    { 
    }
```

aperezdc replied:
Changed, thanks.

carlosgcampos commented on Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp:
```diff
@@ -535,7 +535,7 @@ void AcceleratedBackingStoreDMABuf::didDestroyBuffer(uint64_t id)
 }
 
-void AcceleratedBackingStoreDMABuf::frame(uint64_t bufferID)
+void AcceleratedBackingStoreDMABuf::frame(uint64_t bufferID, const WebCore::Damage&)
```
Message is defined as `std::optional<WebCore::Region>`

aperezdc replied:
Fixed, thanks for noticing.

carlosgcampos reviewed (approved):
Check my comments and fix the build in other ports before landing

aperezdc committed:
[WPE] Generate damage region information
https://bugs.webkit.org/show_bug.cgi?id=265777

Reviewed by Carlos Garcia Campos.

Includes code and ideas from Lauro Moura, Pawel Lampe, Carlos Garcia
Campos, and Miguel Gomez.

Implement tracking which part of a frame change since the previous one.
This may be used to let display systems optimize presentation of the
rendered frames by updating only the parts that have been "damaged".
While usually negligible on desktops, this can result in significant
savings on embedded devices, particularly in memory bandwith and GPU
usage.

Information about damaged frames is collected mainly from repainted
rectangles (from CoordinatedGraphicsLayer) and animated layers (from
TextureMapperLayer), and recorded in a new helper Damage class.

This is a first pass that only propagates the recorded Damage along with
DMA-BUF buffers, and forwards the information to the Wayland compositor
support in WPEPlatform. It should be possible to propagate the
information through EGL where the EGL_{KHR,EXT}_swap_buffers_with_damage
extension is present, which would work when the WPEPlatform API is not
being used; gdk_dmabuf_texture_builder_set_update_region() could be
used for the GTK port; and the optional FB_DAMAGE_CLIPS property for
the DRM platform. These are left to be implemented later on.

The intention for now is to lay out the foundations, and hence the
feature is gated behind a runtime flag. Being an initial implementation,
there are number of shortcomings that may be addressed as follow-ups.

There is an attempt at detecting whether 3D transforms are used, with
the intention of invalidating the whole frame; this mostly works but it
would be better to properly handle these instead of causing full-frame
damage.

There is a pre-existing issue which manifests itself more obviously with
damage tracking enabled: while animating, the renderer content may not
be as crisp as static; typically WebKit would repaint the crisper
version of the content after animations have completed, but if some
parts have not changed since the last frame, they will show up fuzzy
because they are not considered as damaged. This is independent from
damage tracking.

Enabling damage tracking seems add negligible CPU usage: enabling it is
neutral on Speedometer and MotionMark, which for the WPE port are CPU
bound due to rasterization through Cairo being CPU bound.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml: Add feature
  flags to enable damage tracking, and to produce a single rectangle
  with the bounds of the damaged region instead of multiple rectangles.
* Source/WebCore/Headers.cmake:
* Source/WebCore/platform/graphics/Damage.h: Added.
(WebCore::Damage::invalid):
(WebCore::Damage::region const):
(WebCore::Damage::bounds const):
(WebCore::Damage::rects const):
(WebCore::Damage::isEmpty const):
(WebCore::Damage::isInvalid const):
(WebCore::Damage::invalidate):
(WebCore::Damage::add):
(WebCore::Damage::mergeIfNeeded):
(WebCore::Damage::Damage):
(WebCore::operator<<):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::markDamageRectsUnreliable):
* Source/WebCore/platform/graphics/nicosia/NicosiaCompositionLayer.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::TextureMapperLayer):
(WebCore::TextureMapperLayer::paintSelf):
(WebCore::TextureMapperLayer::addChild):
(WebCore::TextureMapperLayer::removeFromParent):
(WebCore::TextureMapperLayer::removeAllChildren):
(WebCore::TextureMapperLayer::applyAnimationsRecursively):
(WebCore::TextureMapperLayer::acceptDamageVisitor):
(WebCore::TextureMapperLayer::dismissDamageVisitor):
(WebCore::TextureMapperLayer::recordDamage):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
(WebCore::TextureMapperLayer::clearDamage):
(WebCore::TextureMapperLayer::invalidateDamage):
(WebCore::TextureMapperLayer::addDamage):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.cpp:
(WebCore::CoordinatedGraphicsLayer::setContentsOpaque):
(WebCore::CoordinatedGraphicsLayer::markDamageRectsUnreliable):
(WebCore::CoordinatedGraphicsLayer::setNeedsDisplay):
(WebCore::CoordinatedGraphicsLayer::setNeedsDisplayInRect):
(WebCore::CoordinatedGraphicsLayer::flushCompositingStateForThisLayerOnly):
(WebCore::CoordinatedGraphicsLayer::computeTransformedVisibleRect):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::layerRendererStyleHas3DTransformOperation):
(WebCore::RenderLayerBacking::updateAfterDescendants):
* Source/WebKit/Shared/CoordinatedGraphics/CoordinatedGraphicsScene.cpp:
(WebKit::CoordinatedGraphicsScene::CoordinatedGraphicsScene):
(WebKit::CoordinatedGraphicsScene::paintToCurrentGLContext):
(WebKit::texmapLayer):
(WebKit::CoordinatedGraphicsScene::updateSceneState):
(WebKit::CoordinatedGraphicsScene::ensureRootLayer):
(WebKit::CoordinatedGraphicsScene::purgeGLResources):
(WebKit::CoordinatedGraphicsScene::recordDamage):
* Source/WebKit/Shared/CoordinatedGraphics/CoordinatedGraphicsScene.h:
(WebKit::CoordinatedGraphicsScene::lastDamage const):
* Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.cpp:
(WebKit::ThreadedCompositor::create):
(WebKit::ThreadedCompositor::ThreadedCompositor):
(WebKit::ThreadedCompositor::renderLayerTree):
* Source/WebKit/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.h:
* Source/WebKit/UIProcess/dmabuf/AcceleratedBackingStoreDMABuf.messages.in:
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp:
(WebKit::AcceleratedBackingStoreDMABuf::frame):
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.h:
* Source/WebKit/UIProcess/wpe/AcceleratedBackingStoreDMABuf.cpp:
(WebKit::AcceleratedBackingStoreDMABuf::frame):
* Source/WebKit/UIProcess/wpe/AcceleratedBackingStoreDMABuf.h:
* Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
(wpe_view_render_buffer):
* Source/WebKit/WPEPlatform/wpe/WPEView.h:
* Source/WebKit/WPEPlatform/wpe/drm/WPEViewDRM.cpp:
(wpeViewDRMRenderBuffer):
* Source/WebKit/WPEPlatform/wpe/headless/WPEViewHeadless.cpp:
(wpeViewHeadlessRenderBuffer):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
* Source/WebKit/WebProcess/WebPage/AcceleratedSurface.h:
(WebKit::AcceleratedSurface::didRenderFrame):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.cpp:
(WebKit::LayerTreeHost::LayerTreeHost):
(WebKit::LayerTreeHost::scrollNonCompositedContents):
(WebKit::LayerTreeHost::didChangeViewport):
(WebKit::LayerTreeHost::didRenderFrame):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.h:
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
(WebKit::AcceleratedSurfaceDMABuf::didRenderFrame):
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.h:
* Source/WebKit/WebProcess/WebPage/libwpe/AcceleratedSurfaceLibWPE.cpp:
(WebKit::AcceleratedSurfaceLibWPE::didRenderFrame):
* Source/WebKit/WebProcess/WebPage/libwpe/AcceleratedSurfaceLibWPE.h:

Canonical link: https://commits.webkit.org/280193@main
